torch/ao/quantization/fx/_decomposed.py

```diff
@@ -0,0 +1,43 @@
+import torch
+from torch.library import Library, impl
+
+quantized_decomposed_lib = Library("quantized_decomposed", "DEF")
+
+quantized_decomposed_lib.define(
+    "quantize_per_tensor(Tensor input, float scale, int zero_point, int quant_min, int quant_max, ScalarType dtype) -> Tensor")
+
+@impl(quantized_decomposed_lib, "quantize_per_tensor", "CompositeExplicitAutograd")
+def quantize_per_tensor(input, scale, zero_point, quant_min, quant_max, dtype):
+    assert input.dtype == torch.float32, f"Expecting input to have dtype torch.float32, but got dtype: {input.dtype}"
+    quant_min_lower_bound = 0
+    quant_max_upper_bound = 0
+    if dtype == torch.uint8:
+        quant_min_lower_bound = 0
+        quant_max_upper_bound = 255
+    elif dtype == torch.int8:
+        quant_min_lower_bound = -128
+        quant_max_upper_bound = 127
+    else:
+        raise ValueError(f"Unsupported dtype: {dtype}")
+
+    assert quant_min >= quant_min_lower_bound, \
+        "quant_min out of bound for dtype, " \
+        f"quant_min_lower_bound: {quant_min_lower_bound} quant_min: {quant_min}"
+
+    assert quant_max <= quant_max_upper_bound, \
+        "quant_max out of bound for dtype, " \
+        f"quant_max_upper_bound: {quant_max_upper_bound} quant_max: {quant_max}"
+
+    inv_scale = 1.0 / scale
+    return torch.clamp(torch.round(input * inv_scale) + zero_point, quant_min, quant_max).to(dtype)
+
+quantized_decomposed_lib.define(
+    "dequantize_per_tensor(Tensor input, float scale, int zero_point, int quant_min, int quant_max, ScalarType dtype) -> Tensor")
+
+@impl(quantized_decomposed_lib, "dequantize_per_tensor", "CompositeExplicitAutograd")
+def dequantize_per_tensor(input, scale, zero_point, quant_min, quant_max, dtype):
```

Comment:
thoughts on why `quant_min` and `quant_max` are in this function?  They are necessary for quantizing, but not sure why they would be needed for dequantizing.

Reply:
I think it might be useful as an metadata for the quantized Tensor input, that could be useful for pattern matching in the future, but I don't have a strong opinion on this one.

Reply:
sorry, what do you mean? not sure I understand yet

Reply:
sure, let me give two examples:

(1). use quant_min/quant_max to show the range restriction of input
pattern for xnnpack linear (simplified):
def forward(self, x):
       x = dequantize(x, scale, zero_point, -128, 127, torch.int8)
       # range restriction for weight
       w = dequantize(x, scale, zero_point, -127, 127, torch.int8)
       x = fp32_linear(x, w, ...)
       x = quantize(x, scale, zero_point, -128, 127, torch.int8)
       return x

(2). using int32 to simulate int16
pattern for a int16 op
def forward(self, x):
    # we use quant_min/quant_max in dequantize to show that we have an int16 input tensor
    x = dequantize(x, scale, zero_point, -2**15, 2**15 - 1, torch.int32)
    x = some_op(x)
    x = quantize(x, scale, zero_point, -2**15, 2**15 - 1, torch.int32)
    return x

Reply:
I understand why quant_min/quant_max are needed for quantize, still don't understand why they are there for dequantize.  Is there some need to use them inside of the dequantize function in the future?  Not sure why their presence or absence would affect pattern matching either.

Reply:
is it just carrying the metadata of "this was quantized with quant_min=a and quant_max=b" around?

Reply:
I think not having quant_min/quant_max in dequantize makes more sense, but don't want to block on that.  The rest of the PR LGTM.

Reply:
> in the new design, quant_min and quant_max might be useful to decide which op the pattern corresponds to, as is shown in the example.

>  I think an alternative can be we traverse back one more step to quantize op to find the quant_min/quant_max metadata for the input of dequantize op,

I understand what you're saying, still missing on why this is needed though.  Could you provide an example where quant_min/quant_max on the quantize is not enough?  I'm not sure that the examples above demonstrate that.

Reply:
Are you trying to say that there will be future logic which would do something like

```
    # we know the int tensor satisfies the constraint of -2**15 <= int_val <= 2**15 - 1, so we can lower to op1
    x = dequantize(x, scale, zero_point, -2**15, 2**15 - 1, torch.int32)
    x = some_op(x)
    x = quantize(x, scale, zero_point, -2**15, 2**15 - 1, torch.int32)

    # we know the int tensor satisfies the constraint of -2**15 <= int_val <= 2**15 - 1, 
    # so we cannot lower to op1 without more information
    x = dequantize(x, scale, zero_point, torch.int32)
    x = some_op(x)
    x = quantize(x, scale, zero_point, -2**15, 2**15 - 1, torch.int32)
```
?

Reply:
Sure, let's say we have the following ops in the same backend:

* op1: xnnpack linear
** activation dtype = int8, range=-128, 127
** weight dtype = int8, range=**-128**, 127
* op2 another linear implementation
** activation dtype = int8, range=-127, 127
** weight dtype = int8, range=**-127**, 127

we'll need the following two patterns to match them:
## op1 pattern:
```
def forward(self, x):
       x = dequantize(x, scale, zero_point, -128, 127, torch.int8)
       # range restriction for weight
       w = dequantize(x, scale, zero_point, -127, 127, torch.int8)
       x = fp32_linear(x, w, ...)
       x = quantize(x, scale, zero_point, -128, 127, torch.int8)
       return x
```

## op2 pattern:
```
def forward(self, x):
       x = dequantize(x, scale, zero_point, -128, 127, torch.int8)
       # full range restriction for weight
       w = dequantize(x, scale, zero_point, -128, 127, torch.int8)
       x = fp32_linear(x, w, ...)
       x = quantize(x, scale, zero_point, -128, 127, torch.int8)
       return x
```

Reply:
yeah, I'm also open to removing it, if there are no such use cases, sg, I'll document this and land for now, we should be able to revisit the decision later if needed, since the reference quantized model are not serialized so we should be able to change that